gdal/doc/source/development/rfc/rfc78_osgeo_utils.rst

```diff
@@ -0,0 +1,280 @@
+.. _rfc-78:
+
+===========================================
+RFC 78: `osgeo_utils` Python package split
+===========================================
+
+============== ============================
+Author:        Idan Miara
+Contact:       idan@miara.com
+Started:       2020-Dec-3
+Last updated:  2020-Dec-7
+Status:        Preliminary, target GDAL 3.3
+============== ============================
+
+Summary
+-------
+
+This RFC is focused on the Python code in the GDAL repo, it would have no impact on non-Python code.
+This RFC suggests that the `osgeo` package will include only the SWIG bindings (As was in GDAL<3.2),
+And mainly suggests to move `osgeo.utils`, `osgeo.auxiliary` and other Python stuff that can be reused
+into a new pure-Python package named `osgeo_utils` and sets a new layout for this package.
+The `osgeo_utils` package will be distributed in pypi.org independently of the `osgeo` package.
```

Comment:
I really don't like those names, at least as pypi packages. For pypi packages, "gdal_bindings" and "gdal_utils" would be better, and "gdal" as the meta.
My feeling is that the ratio complications / advantages of this RFC isn't big enough. And this RFC without the actual pypi packaging splitting (which is where complications would start) doesn't seem to bring much advantages, and I would need to be convinced that there would be many users taking advantages of those split packages.
My vote would be currently somewhere around 0 / -0 on it, besides the naming issue.

Reply:
I am also for splitting the wheels.

So if I understand you correctly you meant:
* `osgeo` in `gdal_bindings` wheel
* `osgeo_utils` in `gdal_utils` wheel, that depends on the `gdal_bindings` wheel
* `gdal` wheel would depend on both as a meta package ?
Does that sound that right?

If wheel/pypi packaging splitting is implemented would you think you'd vote for it?

Reply:
> If wheel/pypi packaging splitting is implemented would you think you'd vote for it?

I don't know. I guess I'm not really convinced this package splitting and all the implications it has, foreseen and unforeseen, is really worth it. I can see your use case, but not sure how common it is, and as discussed privately, you could also fallback to vendorizing the utils you're interested in in your library, in case you depend on new features not typically present in the version of the utility of the GDAL package you rely on.

Reply:
Our maybe we better make `gdal_utils` be also the package name and also the wheel name to avoid further confusion of names.

Reply:
> > If wheel/pypi packaging splitting is implemented would you think you'd vote for it?
> 
> I don't know. I guess I'm not really convinced this package splitting and all the implications it has, foreseen and unforeseen, is really worth it. I can see your use case, but not sure how common it is, and as discussed privately, you could also fallback to vendorizing the utils you're interested in in your library, in case you depend on new features not typically present in the version of the utility of the GDAL package you rely on.

I understand, thank you.

Reply:
As you said my implementation already makes a fairly large diff (without wheel split)
So maybe we better of with a more gradual and conservative two step approach:

1. As already implemented, and wouldn't have a big impact on users who don't care about it as everything would be still in the `gdal` wheel and the utils still installed as scripts for backwards compatibility. Users who care about mixing versions could upgrade the utils with the separate utils wheel (which I also implemented in that PR).
2. Evaluate and implement a wheel split in another PR.

I guess the naming should be decided in this PR so not to add more future confusions.

Reply:
>  Or maybe we better make gdal_utils be also the package name and also the wheel name to avoid further confusion of names.

Would be better

>     1. Users who care about mixing versions could upgrade the utils with the separate utils wheel (which I also implemented in that PR).

Hum, how that work if someone installs the "gdal" all-in-one package and the "gdal_utils" one. Wouldn't that conflict ?

> I guess the naming should be decided in this PR so not to add more future confusions.

Yes, we don't want to change package names for every RFC :-)

Reply:
> Hum, how that work if someone installs the "gdal" all-in-one package and the "gdal_utils" one. Wouldn't that conflict ?

I've done some testing. 

It seems that every time you `pip install` a wheel it just overwrites whichever files that may already exist. 
If you `pip install gdal` then `pip install gdal_utils` you'd get the utils from `gdal_utils`.
If later you do again `pip install gdal` with a different version then you'd get the utils from `gdal` again, and so on.
(it doesn't seem that it matters which version is a bigger number, just which one you installed later)

If you `pip install gdal` then `pip install gdal_utils` and then `pip uninstall gdal_utils` then the `utils` would be uninstalled and you'd be left with gdal without utils.
Then you could `pip install gdal --ignore-installed` to get them back again (`--ignore-installed` is not required if you install a different version)

Reply:
I've been thinking...
`gdal` and `gdal_utils` as wheel/pypi packages sounds good and consistent.
regarding the package inside `gdal` and `gdal_utils`, I'm not sure what sounds more consistent:
option 1:
```
from osgeo import gdal, ogr, osr
from gdal_utils import gdal_calc, ogr_foo, osr_bar
```
option 2:
from osgeo import gdal
from utils_utils import gdal_calc, ogr_foo, osr_bar
```

Regarding the second stage:
I think I'll drop the `gdal` meta package idea. sounds too complicated.
When/if we decide to remove the utils from the `gdal` wheel then they will be already ready inside `gdal_utils` wheel, which depends on the `gdal` wheel anyway.
So `pip install gdal_utils` would install both.

What do you think?